### PR TITLE
[BEAM-13193] Add elements field in ProcessBundleRequest and ProcessBu…

### PR DESCRIPTION
**Pull request by y1chi**
…ndleResponse

**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by y1chi** [comment]
Run Java PreCommit

**Comment by y1chi** [comment]
R: @robertwb @lukecwik

File: model/fn-execution/src/main/proto/beam_fn_api.proto [hunk 1]
```diff
@@ -296,6 +296,14 @@ message ProcessBundleRequest {
...
   // cached data returned by the State API across multiple bundles.
   repeated CacheToken cache_tokens = 2;
+
+  // (optional) Elements to be processed with the bundle. Either all or
+  // none of the bundle elements should be included in the ProcessBundleRequest.
+  // This elements embedding is to achieve better Fn API efficiency for bundles
+  // that contain only small amount of data and are cheap to be processed on the
+  // SDK harness side. See more at
+  // https://s.apache.org/beam-fn-api-control-data-embedding.
+  Elements elements = 3;
```
Comment (lukecwik) [inline model/fn-execution/src/main/proto/beam_fn_api.proto:306]:
```suggestion
  // (optional) Elements to be processed within the bundle. Either all or
  // none of the elements should be included in the ProcessBundleRequest.
  // This embedding is to achieve better efficiency for bundles
  // that contain only small amounts of data. See more at
  // https://s.apache.org/beam-fn-api-control-data-embedding.
  Elements elements = 3;
```

Comment (lukecwik) [reply model/fn-execution/src/main/proto/beam_fn_api.proto:306]: Please add to the comment something like: `This field will can be set if the SDK declares that they support the "...." capability.`

File: model/fn-execution/src/main/proto/beam_fn_api.proto [hunk 2]
```diff
@@ -327,6 +335,13 @@ message ProcessBundleResponse {
   // payload field with the bytes specified here.
   map<string, bytes> monitoring_data = 5;
 
+  // (optional) Output elements of the processed bundle. Either all or
+  // none of the bundle elements should be included in the ProcessBundleResponse.
+  // This elements embedding is to achieve better Fn API efficiency for bundles
+  // that only produce small amount of outputs. See more at
+  // https://s.apache.org/beam-fn-api-control-data-embedding.
```
Comment (lukecwik) [inline model/fn-execution/src/main/proto/beam_fn_api.proto:342]:
```suggestion
  // (optional) Output elements of the processed bundle. Either all or
  // none of the bundle elements should be included in the ProcessBundleResponse.
  // This embedding is to achieve better efficiency for bundles
  // that contain only small amounts of data. See more at
  // https://s.apache.org/beam-fn-api-control-data-embedding.
```

Comment (lukecwik) [reply model/fn-execution/src/main/proto/beam_fn_api.proto:342]: Please add another comment stating that: `This field can only be set if the runner declares that they support the "...." capability.`

File: model/pipeline/src/main/proto/beam_runner_api.proto [hunk 3]
```diff
@@ -1570,6 +1570,12 @@ message StandardProtocols {
     // the entire SDK harness process, not specific to a bundle.
     HARNESS_MONITORING_INFOS = 4
         [(beam_urn) = "beam:protocol:harness_monitoring_infos:v1"];
+
+    // Indicates that this SDK can process elements embedded in Fn API
+    // control plane ProcessBundleRequest. See more about the protocol at
+    // https://s.apache.org/beam-fn-api-control-data-embedding
```
Comment (lukecwik) [inline model/pipeline/src/main/proto/beam_runner_api.proto:1576]:
```suggestion
    // Indicates that this SDK can process elements embedded in the
    // ProcessBundleRequest. See more about the protocol at
    // https://s.apache.org/beam-fn-api-control-data-embedding
```

File: model/pipeline/src/main/proto/beam_runner_api.proto [hunk 4]
```diff
@@ -1579,6 +1585,12 @@ message StandardRunnerProtocols {
   enum Enum {
     // Indicates suport the MonitoringInfo short id protocol.
     MONITORING_INFO_SHORT_IDS = 0 [(beam_urn) = "beam:protocol:monitoring_info_short_ids:v1"];
+
+    // Indicates that this runner can process elements embedded in Fn API
+    // control plane ProcessBundleResponse. See more about the protocol at
+    // https://s.apache.org/beam-fn-api-control-data-embedding
```
Comment (lukecwik) [inline model/pipeline/src/main/proto/beam_runner_api.proto:1591]:
```suggestion
    // Indicates that this runner can process elements embedded in the
    // ProcessBundleResponse. See more about the protocol at
    // https://s.apache.org/beam-fn-api-control-data-embedding
```

File: model/fn-execution/src/main/proto/beam_fn_api.proto (hunk 1 above)
Comment (y1chi) [reply model/fn-execution/src/main/proto/beam_fn_api.proto]: Done.

File: model/fn-execution/src/main/proto/beam_fn_api.proto (hunk 2 above)
Comment (y1chi) [reply model/fn-execution/src/main/proto/beam_fn_api.proto]: Done.

File: model/pipeline/src/main/proto/beam_runner_api.proto (hunk 4 above)
Comment (y1chi) [reply model/pipeline/src/main/proto/beam_runner_api.proto]: Done.

File: model/pipeline/src/main/proto/beam_runner_api.proto (hunk 3 above)
Comment (y1chi) [reply model/pipeline/src/main/proto/beam_runner_api.proto]: Done.

File: model/fn-execution/src/main/proto/beam_fn_api.proto (hunk 1 above)
Comment (lukecwik) [reply model/fn-execution/src/main/proto/beam_fn_api.proto]: What do you think of my first comment which updated the existing text as well?

File: model/fn-execution/src/main/proto/beam_fn_api.proto (hunk 2 above)
Comment (lukecwik) [reply model/fn-execution/src/main/proto/beam_fn_api.proto]: What do you think of my first comment which updated the existing text as well?

File: model/fn-execution/src/main/proto/beam_fn_api.proto [hunk 5]
```diff
@@ -296,6 +296,15 @@ message ProcessBundleRequest {
   // (Optional) A list of cache tokens that can be used by an SDK to reuse
   // cached data returned by the State API across multiple bundles.
   repeated CacheToken cache_tokens = 2;
+
+  // (optional) Elements to be processed with the bundle. Either all or
+  // none of the bundle elements should be included in the ProcessBundleRequest.
+  // This elements embedding is to achieve better Fn API efficiency for bundles
+  // that contain only small amount of data and are cheap to be processed on the
+  // SDK harness side. This field can be set only if the SDK declares that it
+  // supports the CONTROL_REQUEST_ELEMENTS_EMBEDDING capability. See more at
```
Comment (lukecwik) [inline model/fn-execution/src/main/proto/beam_fn_api.proto:305]:
```suggestion
  // supports the beam:protocol:control_request_elements_embedding:v1 capability. See more at
```

Note that the ENUM can change but its contents of string can't so best to use it within the comment.

File: model/fn-execution/src/main/proto/beam_fn_api.proto [hunk 6]
```diff
@@ -327,6 +336,15 @@ message ProcessBundleResponse {
   // payload field with the bytes specified here.
   map<string, bytes> monitoring_data = 5;
 
+  // (optional) Output elements of the processed bundle. Either all or
+  // none of the bundle elements should be included in the ProcessBundleResponse.
+  // This embedding is to achieve better efficiency for bundles that only
+  // contain small amounts of data. his field can be set only if the runner
+  // declares that it supports the CONTROL_RESPONSE_ELEMENTS_EMBEDDING
```
Comment (lukecwik) [inline model/fn-execution/src/main/proto/beam_fn_api.proto:343]:
```suggestion
  // declares that it supports the beam:protocol:control_response_elements_embedding:v1
```

Note that the ENUM can change but its contents of string can't so best to use it within the comment.

File: model/fn-execution/src/main/proto/beam_fn_api.proto (hunk 1 above)
Comment (y1chi) [reply model/fn-execution/src/main/proto/beam_fn_api.proto]: I guess I missed one of the change and thought I've updated both.

**Review by lukecwik — APPROVED** [review]

File: model/fn-execution/src/main/proto/beam_fn_api.proto (hunk 5 above)
Comment (y1chi) [reply model/fn-execution/src/main/proto/beam_fn_api.proto]: Done.

File: model/fn-execution/src/main/proto/beam_fn_api.proto (hunk 6 above)
Comment (y1chi) [reply model/fn-execution/src/main/proto/beam_fn_api.proto]: Done.

**Comment by y1chi** [comment]
Run Java PreCommit

**Comment by y1chi** [comment]
Run Python_PVR_Flink PreCommit

**Comment by y1chi** [comment]
Run Java PreCommit

**Pull request merged**